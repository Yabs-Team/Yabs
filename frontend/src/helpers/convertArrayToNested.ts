export default function convertList(payload: any, typeOfId: string) {
  return payload.reduce((acc: any, item: any) => {
    acc[item[typeOfId]] = item;
    return acc;
  }, {});
}
