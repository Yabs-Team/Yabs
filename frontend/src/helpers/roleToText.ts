export default function roleToText(role: number): string {
  const roleText = {
    1: 'Elev',
    2: 'Elevhälsoteam',
    4: 'Lärare',
    16: 'Admin',
    32: 'Rektor'
  };
  let text = roleText[role];
  if(!text) {
    text = 'Role: ' + role;
  }
  return text;
}