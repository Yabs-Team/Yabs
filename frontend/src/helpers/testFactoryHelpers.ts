import {
  mount,
  shallowMount,
  VueClass,
  Wrapper,
  ThisTypedMountOptions,
  createLocalVue,
} from '@vue/test-utils';
import Vue from 'vue';
import BootstrapVue from 'bootstrap-vue';
import Vuex from 'vuex';
import {cloneDeep} from 'lodash';
import users from '@/store/modules/users';

export function factory<V extends Vue>(
  component: VueClass<V>,
  options?: ThisTypedMountOptions<V>,
  mountFunc: any = mount,
): Wrapper<V> {
  const localVue = createLocalVue();

  localVue.use(BootstrapVue);
  localVue.use(Vuex);

  const storeConfig = {
    modules: {
      users,
    },
  };

  const store = new Vuex.Store(cloneDeep(storeConfig));

  return mountFunc(component, {
    localVue,
    store,
    ...options,
  });
}

// my jokes aren't dry unlike my code...

export function shallowFactory<V extends Vue>(
  component: VueClass<V>,
  options?: ThisTypedMountOptions<V>,
): Wrapper<V> {
  return factory(component, options, shallowMount);
}

export type IRouterFactory<V extends Vue> = [jest.Mock<any, any>, Wrapper<V>];
export function routerFactory<V extends Vue>(
  component: VueClass<V>,
  options?: ThisTypedMountOptions<V>,
): IRouterFactory<V> {
  const pushFunction = jest.fn();

  return [
    pushFunction,
    factory(component, {
      mocks: {
        $router: {
          push: pushFunction,
        },
      },
      ...options,
    }),
  ];
}

export function shallowRouterFactory<V extends Vue>(
  component: VueClass<V>,
  options?: ThisTypedMountOptions<V>,
): IRouterFactory<V> {
  const pushFunction = jest.fn();

  return [
    pushFunction,
    shallowFactory(component, {
      mocks: {
        $router: {
          push: pushFunction,
        },
      },
      ...options,
    }),
  ];
}
