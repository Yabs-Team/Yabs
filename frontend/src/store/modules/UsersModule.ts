import Vue from 'vue';
import {
  VuexModule,
  Module,
  getModule,
  Action,
  Mutation,
} from 'vuex-module-decorators';
import store from '..';
import UsersAPI from '../../api/users';
import convertList from '../../helpers/convertArrayToNested';
export interface User {
  created_at: string;
  email: string;
  google_token: string;
  klass: string;
  name: string;
  photo_path: string;
  role: string;
  uid: number;
  updated_at: string;
}

export interface UserCollection { [uid: number]: User; }

export interface UserState {
  users: UserCollection;
  currentUser: any;
  failure: any;
}

@Module({dynamic: true, namespaced: true, name: 'UsersModule', store, preserveState: true})
class UsersModule extends VuexModule {
  private userState: UserState = {users: {}, currentUser: null, failure: null};

  get currentUserID(): number {
    return this.userState.currentUser;
  }

  get all(): UserCollection {
    return this.userState.users;
  }

  get currentUser(): User {
    return this.all[this.userState.currentUser];
  }

  @Action({rawError: true})
  public fetchAll(): Promise<object> {
    return new Promise((resolve, reject) => {
      UsersAPI.all()
        .then((response: User[]) => {
          this.convertUserList(response);
          resolve();
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public update(request: any) {
    return new Promise((resolve, reject) => {
      UsersAPI.update(request)
        .then((response: any) => {
          this.setUser(response);
          this.setCurrentUser(response);
          resolve(response);
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public signIn(request: any) {
    return new Promise((resolve, reject) => {
      UsersAPI.signIn(request)
        .then((response: any) => {
          this.setCurrentUser(response);
          resolve(response);
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public signOut(): Promise<object> {
    return new Promise((resolve, reject) => {
      UsersAPI.signOut()
        .then((response: any) => {
          this.setCurrentUser(response);
          resolve(response);
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Mutation
  public setUser(payload: any): void {
    Vue.set(this.userState.users, payload.uid, payload);
  }

  @Mutation
  public setCurrentUser(payload: any): void {
    Vue.set(this.userState, 'currentUser', payload.uid);
  }

  @Mutation
  public convertUserList(payload: User[]): void {
    const list = convertList(payload, 'uid');
    Vue.set(this.userState, 'users', list);
  }
}

export default getModule(UsersModule);
