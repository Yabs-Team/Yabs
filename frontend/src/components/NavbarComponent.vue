<template>
  <div class="pb-10">
    <v-app-bar
      fixed
      class="nav"
    >
      <v-toolbar-items>
        <v-btn
          to="/"
          text
        >
          <v-icon class="px-3">
            mdi-home
          </v-icon>Yabs
        </v-btn>
      </v-toolbar-items>
      <v-spacer />
      <v-toolbar-items v-if="UsersModule.currentUser">
        <v-btn
          :to="'/users/'+ UsersModule.currentUserID"
          text
        >
          <v-icon class="px-3">
            account_circle
          </v-icon>Profil
        </v-btn>
        <v-btn
          to="/admin"
          text
        >
          <v-icon class="px-3">
            add_circle
          </v-icon>Admin
        </v-btn>
        <v-btn
          class="ml-2 mr-5"
          text
          color="grey"
          @click="signOut"
        >
          <v-icon>mdi-logout-variant</v-icon>
        </v-btn>
      </v-toolbar-items>
      <v-toolbar-items v-else>
        <GoogleLogin />
      </v-toolbar-items>
    </v-app-bar>
  </div>
</template>

<script lang="ts">
import GoogleLogin from '@/components/GoogleLogin.vue';
import UsersModule from '../store/modules/UsersModule';
import { defineComponent } from '@vue/composition-api';

export default defineComponent({
  name: 'Navbar',
  components: { GoogleLoginÂ },
  setup(){
    async function signOut() : Promise<void> {
      try {
        // @ts-ignore: gapi
        const auth2 = await gapi.auth2.getAuthInstance();
        await auth2.signOut();
        await UsersModule.signOut();
      } catch (e) {
        UsersModule.signOut();
      }
    }
    return{
      signOut, UsersModule
    };
  }
});
</script>

<style>
a:hover {
  text-decoration: none !important;
}
a:active,
a:focus {
  outline: 0;
  border: none;
  -moz-outline-style: none;
  outline-style: none;
}
button:active,
button:focus {
  outline: 0 !important;
  border: none !important;
  -moz-outline-style: none;
  outline-style: none;
}

input:-webkit-autofill {
  background-color: transparent !important;
  background-image: none !important;
  color: #000000 !important;
}

input:-webkit-autofill {
  -webkit-box-shadow: 200px 200px 100px white inset;
  box-shadow: 200px 200px 100px white inset;
}
</style>

