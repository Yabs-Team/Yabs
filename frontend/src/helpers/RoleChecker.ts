import UsersModule from '../store/modules/UsersModule';
import { User } from '@/types';

interface Roles { [bitValue: number]: string}

const roles: Roles = {
  1: 'Elev',
  2: 'Elevhälsoteam',
  4: 'Lärare',
  8: 'Vaktmästare',
  16: 'Administratör',
  32: 'Rektor'
};

export default class RoleChecker {

  static checkPermission(required: number): boolean {
    const user: User = UsersModule.currentUser;
    return !!user && ((user.role & required) == required || (user.role > required));
  }
  
  static isPrincipal(): boolean {
    return this.checkPermission(32);
  }

  static isAdmin(): boolean {
    return this.checkPermission(16);
  }

  static isCaretaker(): boolean {
    return this.checkPermission(8);
  }

  static isTeacher(): boolean {
    return this.checkPermission(4);
  }

  static isStudentHealth(): boolean {
    return this.checkPermission(2);
  }

  static isStudent(): boolean {
    return UsersModule.currentUser && (UsersModule.currentUser.role & 1) == 1;
  }

  static roleAsText(user: User = UsersModule.currentUser): string {
    if (!user) {return ''; }
    for(const bitValue of Object.keys(roles).reverse()) {
      if (user.role >= Number(bitValue)) {
        return roles[Number(bitValue)];
      }
    }
    return '';
  }
}
