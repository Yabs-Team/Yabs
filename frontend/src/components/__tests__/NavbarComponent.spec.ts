import NavbarComponent from '@/components/NavbarComponent.vue';

import { Wrapper } from '@vue/test-utils';
import { shallowFactory } from '@/helpers/testFactoryHelpers';
import UsersModule from '@/store/modules/UsersModule';

describe('NavbarComponent.vue', () => {

  const injectRouterStubs = {stubs: ['router-link', 'router-view']};

  it('renders title when passed', () => {
    const wrapper: Wrapper<NavbarComponent> = shallowFactory(NavbarComponent, injectRouterStubs);

    expect(wrapper.html()).toMatch('Yabs');
  });


  it('renders correctly', () => {
    const wrapper: Wrapper<NavbarComponent> = shallowFactory(NavbarComponent, injectRouterStubs);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
