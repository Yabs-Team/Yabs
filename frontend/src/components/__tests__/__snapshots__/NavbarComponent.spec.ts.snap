// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarComponent.vue renders correctly 1`] = `
<div class="base">
  <b-navbar-stub tag="nav" type="dark" variant="primary" toggleable="true">
    <b-navbar-toggle-stub label="Toggle navigation" target="nav_dropdown_collapse"></b-navbar-toggle-stub>
    <b-collapse-stub id="nav_dropdown_collapse" isnav="true" tag="div">
      <b-navbar-nav-stub tag="ul">
        <b-nav-item-stub target="_self" event="click" routertag="a">
          <router-link-stub to="/" class="link">
            Start
          </router-link-stub>
        </b-nav-item-stub>
        <b-nav-item-stub target="_self" event="click" routertag="a">
          <router-link-stub to="/users/undefined" class="link">
            Profil
          </router-link-stub>
        </b-nav-item-stub>
        <b-nav-item-stub target="_self" event="click" routertag="a">
          <router-link-stub to="/add" class="link">
            LÃ¤gg till
          </router-link-stub>
        </b-nav-item-stub>
        <b-nav-item-stub target="_self" event="click" routertag="a">
          <router-link-stub to="/find" class="link">
            Hitta
          </router-link-stub>
        </b-nav-item-stub>
      </b-navbar-nav-stub>
    </b-collapse-stub>
    <b-navbar-brand-stub target="_self" event="click" routertag="a" tag="div">Yabs</b-navbar-brand-stub>
    <googlelogin-stub></googlelogin-stub>
  </b-navbar-stub>
</div>
`;
